### DISCUSSION TITLE
Incorrect colour formats when using PIL image

### DISCUSSION THREAD

**Discussion by MikeLerwill99** [raspberrypi/picamera2#568] [Q&A]
If I use 
            image = picam2.capture_image ("main")

I always end up with an image that has the red and blue the wrong way round irrespective of whether I specify RGB888 or BGR888 for the format.

However if I instead do
            array = picam2.capture_array ("main")
            image = Image.fromarray(array, mode = "RGB")

Then the colours in the image are correct when i use RGB888 and red and blue swapped as expected if I use BGR888

Can anyone help with why the first does not work correctly as I would rather avoid the double handling of the data.

Mike Lerwill

**Comment by sandyol55** [comment]
What are you using to 'show' the images?

The inbuilt PIL .show method should give a correctly ordered representation, but as an example OpenCv has a reversed default representation.


```
#!/usr/bin/python3
import time
import cv2
import numpy as np
from picamera2 import Picamera2

picam2 = Picamera2()
picam2.start()
time.sleep(1)

image = picam2.capture_image("main")
image.show()

cv2_img=np.array(image)
cv2.imshow("Window",cv2_img)
  
cv2.waitKey(0)
time.sleep(1)

picam2.close()
cv2.destroyAllWindows()
```

**Reply by MikeLerwill99** [reply]
I see I had not provided enough data, 
after image = picam2.capture_image ("main")
I was then
            image = image.resize((720,720))
            array = image.tobytes()
and transferring the resultant array over to another computer for further processing.
The key thing being that I was getting the same data in "array" regardless of the specified format RGB888 or BGR888.

I am no longer using the image intermediate representation so do not have this problem at present

Thanks for your help anyway
regards
Mike

**Comment by sandyol55** [comment]
Ok. Glad you've got a way around it.
It **may** be due to capture_image method defaulting to one format, irrespective of the configuration format??
(Not sure of that though!)

The amended code below captures direct to an array, rather than to an image and then converting to an array, and does seem to respect the alternative RGB888 and BGR888 configurations


```
import time
import cv2
import numpy as np
from picamera2 import Picamera2

picam2 = Picamera2()
still_config=picam2.create_still_configuration(main={"size":(720,720),"format":"RGB888"})
#still_config=picam2.create_still_configuration(main={"size":(720,720),"format":"BGR888"})
picam2.configure(still_config)

picam2.start()
time.sleep(1)

image = picam2.capture_array("main")

cv2_img=np.array(image)
cv2.imshow("Window",cv2_img)
  
cv2.waitKey(0)
time.sleep(1)

picam2.close()
cv2.destroyAllWindows()
```

Regards
Sandy
